File: kernel/drivers/gpio.c
#include "drivers/base.h"
#include "drivers/gpio.h"
#include "util/stdlib.h"

/* All the magic here is described in Section 6 */
// QUESTION: What's the point of GPIO pins above 27?

#define GPBASE (PBASE + 0x00200000)

#define FSEL_REG(x) (GPBASE + (x)*0x4)

/*
  Set GPIO behavior, either input, output, or an alterative function.
*/
int gpio_set_func(uint8_t pin, gpfunc_t f) {
  // Invalid pin
  if (pin < 2 || pin > 27) {
    return 1;
  }

  // Get register
  uint64_t REG = FSEL_REG(pin / 10);
  uint32_t cur = get32(REG);
  // Each pin takes a 3bit func code
  int shift = 3 * (pin % 10);
  // Clear pin
  cur &= ~(0b111 << shift);
  // Set pin
  cur |= (f << shift);

  put32(REG, cur);

  return 0;
}

#define GPPUD (GPBASE + 0x94)
#define GPPUDCLK0 (GPBASE + 0x98)
#define GPPUDCLK1 (GPBASE + 0x9C) // Not used

int gpio_set_pull(uint8_t pin, gppull_t pull) {
  // Invalid pin
  if (pin < 2 || pin > 27) {
    return 1;
  }

  // Set pull
  put32(GPPUD, pull);
  delay(150);
  // Set clock
  put32(GPPUDCLK0, 1 << pin);
  delay(150);
  // Clear clock
  put32(GPPUD, 0);
  put32(GPPUDCLK0, 0);

  return 0;
}

#define GPSET0 (GPBASE + 0x1C)
#define GPSET1 (GPBASE + 0x20) // Not used
#define GPCLR0 (GPBASE + 0x28)
#define GPCLR1 (GPBASE + 0x2C) // Not used

int gpio_set(uint8_t pin, int val) {
  // Invalid pin
  if (pin < 2 || pin > 27) {
    return 1;
  }

  uint64_t REG = val ? GPSET0 : GPCLR0;
  put32(REG, 1 << pin);

  return 0;
}

File: kernel/src/drivers/gpio.rs
use drivers::time;

const GPIO_BASE: usize = super::P_BASE + 0x0020_0000;

const fn fsel_reg(x: u8) -> *mut u32 {
    (GPIO_BASE + (x as usize) * 0x4) as *mut u32
}

#[inline]
fn valid_pin(pin: u8) -> bool {
    2 <= pin && pin <= 27
}

pub enum PinFunc {
    Input = 0,
    Output = 1,
    F5 = 2,
    F4 = 3,
    F0 = 4,
    F1 = 5,
    F2 = 6,
    F3 = 7,
}

pub fn set_func(pin: u8, f: PinFunc) {
    assert!(valid_pin(pin));

    let reg = fsel_reg(pin / 10);
    let shift = 3 * (pin % 10);

    unsafe {
        let mut cur = reg.read_volatile();
        cur &= !(0b111 << shift);
        cur |= (f as u32) << shift;
        reg.write_volatile(cur);
    }
}

#[derive(Copy, Clone)]
pub enum PinPull {
    None = 0,
    Down = 1,
    Up = 2,
}

const GPPUD: *mut u32 = (GPIO_BASE + 0x94) as *mut u32;
const GPPUDCLK0: *mut u32 = (GPIO_BASE + 0x98) as *mut u32;
const GPPUDCLK1: *mut u32 = (GPIO_BASE + 0x9C) as *mut u32;

pub fn set_pull(pin: u8, pull: PinPull) {
    assert!(valid_pin(pin));

    unsafe {
        // Set pull
        GPPUD.write_volatile(pull as u32);
        time::spin_sleep_ins(150);

        // Set Clock
        GPPUDCLK0.write_volatile(pull as u32);
        time::spin_sleep_ins(150);

        // Clear Clock
        GPPUD.write_volatile(0);
        GPPUDCLK0.write_volatile(0);
    }
}

const GPSET0: *mut u32 = (GPIO_BASE + 0x1C) as *mut u32;
const GPSET1: *mut u32 = (GPIO_BASE + 0x20) as *mut u32;
const GPCLR0: *mut u32 = (GPIO_BASE + 0x28) as *mut u32;
const GPCLR1: *mut u32 = (GPIO_BASE + 0x2C) as *mut u32;

pub fn set(pin: u8, val: bool) {
    assert!(valid_pin(pin));

    let reg = if val { GPSET0 } else { GPCLR0 };
    unsafe {
        reg.write_volatile(1 << pin);
    }
}
